agents/aggregator/aggregator_agent.py

```diff
@@ -18,463 +13,511 @@ class Provider:
+    G3 module which handles file rotation.
+    After time_per_file has elapsed, the rotator will end that file and create
+    a new file with the `filename` function.
+    It will write the last_session and last_status frame to any new file if they
+    exist.
+
+    Args:
+        time_per_file (float):
+            time (seconds) before a new file should be written
+        filename (callable):
+            function that generates new filenames.
+    """
+    def __init__(self, time_per_file, filename):
+        self.time_per_file = time_per_file
+        self.filename = filename
+
+        self.file_start_time = None
+        self.writer = None
+        self.last_session = None
+        self.last_status = None
+
+    def close_file(self):
+        self.writer(core.G3Frame(core.G3FrameType.EndProcessing))
+        self.writer = None
+
+    def flush(self):
+        """Flushes current g3 file to disk"""
+        if self.writer is not None:
+            self.writer.Flush()
+
+    def Process(self, frame):
+        """
+        Writes frame to current file. If file has not been started
+        or time_per_file has elapsed, file is closed and a new file is created
+        by `filename` function passed to constructor
         """
-        if hksess is not None:
-            frame = hksess.data_frame(prov_id=self.prov_id)
-        else:
-            frame = core.G3Frame(core.G3FrameType.Housekeeping)
+        ftype = frame['hkagg_type']
 
-        frame['address'] = self.address
-        frame['provider_session_id'] = self.session_id
+        if ftype == so3g.HKFrameType.session:
+            self.last_session = frame
+        elif ftype == so3g.HKFrameType.status:
+            self.last_status = frame
 
-        for block_name, block in self.blocks.items():
-            if not block.timestamps:
-                continue
+        if self.writer is None:
+            self.writer = core.G3Writer(self.filename())
+            self.file_start_time = time.time()
+
+            if ftype in [so3g.HKFrameType.data, so3g.HKFrameType.status]:
+                if self.last_session is not None:
+                    self.writer(self.last_session)
 
-            hk = so3g.IrregBlockDouble()
-            hk.prefix = block.prefix
-            hk.t = block.timestamps
-            for key, ts in block.data.items():
-                hk.data[key] = ts
+            if ftype == so3g.HKFrameType.data:
+                if self.last_status is not None:
+                    self.writer(self.last_status)
 
-            frame['blocks'].append(hk)
+        self.writer(frame)
 
-        return frame
+        if (time.time() - self.file_start_time) > self.time_per_file:
+            self.close_file()
 
 
-class DataAggregator:
+class Aggregator:
     """
-    The Data Aggregator Agent listens to data provider feeds, and outputs the housekeeping data as G3Frames.
+    The Data Aggregator Agent listens to data provider feeds, and outputs the
+    housekeeping data as G3Frames.
```

Comment:
Would it be less awkward if instead I made all the non-thread-safe variables (pids, providers, frame_queue, ...) local to the run function, and passed as arguments to the functions? Then there would be no question about if it is unsafe to call `add_provider`, but it would make it more awkward to call, since you'd have to pass in the `providers` and `pids` objects.

Reply:
Or maybe all of those objects and functions that should be put into a ProviderCollection class, that is instantiated in the run thread... That seems like the best way of doing it.